docs/usage/shoot_pod_autoscaling_best_practices.md

```diff
@@ -0,0 +1,327 @@
+  Therefore, if you do not know the healthy range, do not set CPU limits. If you as author of the source code know its healthy range, set them to the upper threshold of that healthy range (everything above, from your knowledge of that code, is definitely an unbound busy loop or similar, which is the main reason for CPU limits, besides batch jobs where throttling is acceptable or even desired).
+  - Memory limits may be more useful, but suffer a similar, though not as negative downside. As with CPU limits, memory limits aren't helpful on an under-utilized node (=may result in needless outages), but different than CPU limits, they result in an OOM, which triggers VPA to provide more memory suddenly (modifies the currently computed recommendations by a configurable factor, defaulting to +20%, see [docs](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/README.md#custom-memory-bump-up-after-oomkill)).<br/>
+  Therefore, if you do not know the healthy range, do not set memory limits. If you as author of the source code know its healthy range, set them to the upper threshold of that healthy range (everything above, from your knowledge of that code, is definitely an unbound memory leak or similar, which is the main reason for memory limits)
+- Horizontal Pod Autoscaling (HPA): Use always for pods that support horizontal scaling. Prefer scaling on usage, not utilization, as this is more predictable (not dependent on a second variable, namely the current requests) and conflict-free with vertical pod autoscaling (VPA).
+- As a rule of thumb, set the initial replicas to the 5th percentile of the actually observed replica count in production. Since HPA reacts fast, this is not as crucial, but may help reduce initial load on the control plane early after deployment. However, be cautious when you update the higher-level resource not to inadvertently reset the current HPA-controlled replica count (very easy to make mistake that can lead to catastrophic loss of pods). HPA modifies the replica count directly in the spec and you do not want to overwrite that. Even if it reacts fast, it is not instant (not via a mutating webhook as VPA operates) and the damage may already be done.
+- As for minimum and maximum, let your high availability requirements determine the minimum and your theoretical maximum load determine the maximum, flanked with alerts to detect erroneous run-away out-scaling or the actual nearing of your practical maximum load, so that you can intervene.
+- Vertical Pod Autoscaling (VPA): Use always for containers that have a significant usage (e.g. any container above 50m CPU or 100M memory) and a significant spread over time (by more than 2x), i.e. ignore small (e.g. side-cars) or static (e.g. Java statically allocated heap) containers, but otherwise use it to provide the resources needed on the one hand and keep the costs in check on the other hand.
```

Comment:
Hmm... this sounds very surprising to me @andrerun. Why do you say that? If anything, it's the other way around, is it not? Proof of that is how we deploy our own workload. In almost no cases HPA applies. In almost all cases VPA not only applies, but is absolutely and critically necessary as we have seen from countless outages: at first we had nothing under VPA and then more and more and mostly we added yet another VPA when we had another outage for a component that had no VPA before.

Reply:
@andrerun I removed the word "always" for HPA and VPA and maybe that's a compromise.
However, I do not intend to write what you suggested:

> The default choice is horizontal scaling. Only when horizontal scaling does not work (well) for the particular workload, is vertical scaling considered.

This is nothing I can subscribe to/goes against what I saw/see/think. In my world, HPA is useful only for very special workload, but in most cases, VPA is more relevant and often even necessary as we control less and less the software we deploy - we cannot scale it statically, because we can no longer predict its resource consumption with 100s of sub components that we haven't developed ourselves. Back in the day when I did assembler or C development, I knew exactly what my memory consumption was, but that is over since a very long time as software got more and more complex, built from libraries and packages I know only the APIs of.